pallets/gas/src/lib.rs

```diff
@@ -96,28 +98,82 @@ impl ValueNode {
         self.spec_refs.saturating_add(self.unspec_refs)
     }
 
-    /// The first upstream node (self included), that holds a concrete value.
+    /// The first upstream node (self included), that is able to hold a concrete value, but doesn't
+    /// necessarily has a non-zero value.
     /// If some node along the upstream path is missing, returns an error (tree is invalidated).
     pub fn node_with_value<T: Config>(&self) -> Result<ValueNode, DispatchError> {
-        if let ValueType::UnspecifiedLocal { parent } = self.inner {
-            <Pallet<T>>::get_node(parent)
-                .ok_or(Error::<T>::GasTreeInvalidated)?
-                .node_with_value::<T>()
-        } else {
-            Ok(self.clone())
+        let mut ret_node = self.clone();
+        while let ValueType::UnspecifiedLocal { parent } = ret_node.inner {
+            ret_node = <Pallet<T>>::get_node(parent).ok_or(Error::<T>::GasTreeInvalidated)?;
         }
+
+        Ok(ret_node)
     }
 
-    /// Returns root node of the value tree where caller node exists.
+    /// Returns the AccountId (as Origin) of the value tree creator.
+    /// If some node along the upstream path is missing, returns an error (tree is invalidated).
     pub fn root<T: Config>(&self) -> Result<Self, DispatchError> {
-        match self.inner {
-            ValueType::External { .. } => Ok(self.clone()),
-            ValueType::SpecifiedLocal { parent, .. } | ValueType::UnspecifiedLocal { parent } => {
-                <Pallet<T>>::get_node(parent)
-                    .ok_or(<Error<T>>::GasTreeInvalidated)?
-                    .root::<T>()
+        let mut ret_node = self.clone();
+        while let Some(parent) = ret_node.parent() {
+            ret_node = <Pallet<T>>::get_node(parent).ok_or(Error::<T>::GasTreeInvalidated)?;
+        }
+        Ok(ret_node)
+    }
+
+    fn decrease_parents_ref<T: Config>(&self) -> DispatchResult {
+        if let Some(id) = self.parent() {
+            let mut parent = <Pallet<T>>::get_node(id).ok_or(Error::<T>::GasTreeInvalidated)?;
+            assert!(
+                parent.refs() > 0,
+                "parent node must contain ref to its child node"
+            );
+
+            if let ValueType::SpecifiedLocal { .. } = self.inner {
+                parent.spec_refs = parent.spec_refs.saturating_sub(1);
+            } else {
+                parent.unspec_refs = parent.unspec_refs.saturating_sub(1);
+            }
```

Comment:
I didn't get your intention. This code is run in the block, which guarantees, that node has a parent. It means, it can't be an external type node

Reply:
or you mean something like this instead of the selected part
```rust
            match self.inner {
                ValueType::SpecifiedLocal { .. } => {
                    parent.spec_refs = parent.spec_refs.saturating_sub(1)
                }
                ValueType::UnspecifiedLocal { .. } => {
                    parent.unspec_refs = parent.unspec_refs.saturating_sub(1)
                }
                ValueType::External { .. } => {
                    unreachable!("node is guaranteed to have a parent, so can't be an external one")
                }
            }
```